backend/internal/oauth/oauth2/granthandlers/tokenexchange.go

```diff
@@ -0,0 +1,432 @@
+	}
+
+	// If subject token has no scopes, reject requests asking for scopes
+	if len(subjectScopes) == 0 {
+		return nil, &model.ErrorResponse{
+			Error:            constants.ErrorInvalidScope,
+			ErrorDescription: "Cannot request scopes when subject token has no scopes",
+		}
+	}
+
+	// Validate requested scopes are subset of subject scopes
+	subjectScopeSet := make(map[string]bool)
+	for _, s := range subjectScopes {
+		subjectScopeSet[s] = true
+	}
+
+	for _, requestedScope := range requestedScopes {
+		if !subjectScopeSet[requestedScope] {
+			return nil, &model.ErrorResponse{
+				Error: constants.ErrorInvalidScope,
+				ErrorDescription: fmt.Sprintf(
+					"Requested scope '%s' not present in subject token",
+					requestedScope,
+				),
+			}
+		}
+	}
+
+	return requestedScopes, nil
+}
+
+// getAudience determines the audience for the new token.
+func (h *tokenExchangeGrantHandler) getAudience(
+	tokenRequest *model.TokenRequest,
+	subjectClaims map[string]interface{},
+) string {
+	if tokenRequest.Audience != "" {
+		return tokenRequest.Audience
+	}
+
+	if tokenRequest.Resource != "" {
+		return tokenRequest.Resource
+	}
+
+	if aud, ok := subjectClaims["aud"].(string); ok && aud != "" {
+		return aud
+	}
+
+	return tokenRequest.ClientID
+}
+
+// buildTokenClaims constructs the claims for the new token.
+func (h *tokenExchangeGrantHandler) buildTokenClaims(
+	tokenRequest *model.TokenRequest,
+	actorClaims map[string]interface{},
+	subjectActor map[string]interface{},
+	userAttributes map[string]interface{},
+	scopes []string,
+) map[string]interface{} {
+	newClaims := make(map[string]interface{})
+
+	if len(scopes) > 0 {
+		newClaims["scope"] = strings.Join(scopes, " ")
+	}
+
+	newClaims["client_id"] = tokenRequest.ClientID
+
+	if actorClaims != nil {
+		actClaim := map[string]interface{}{}
+		if sub, ok := actorClaims["sub"].(string); ok {
+			actClaim["sub"] = sub
+		}
+		if iss, ok := actorClaims["iss"].(string); ok && iss != "" {
+			actClaim["iss"] = iss
+		}
+
+		// If actor token has its own act claim, preserve it in the chain
+		if actorAct, ok := actorClaims["act"].(map[string]interface{}); ok {
+			if subjectActor != nil {
+				nestedAct := map[string]interface{}{}
+				for k, v := range actorAct {
+					nestedAct[k] = v
+				}
+				nestedAct["act"] = subjectActor
+				actClaim["act"] = nestedAct
+			} else {
+				actClaim["act"] = actorAct
+			}
+		} else if subjectActor != nil {
+			actClaim["act"] = subjectActor
+		}
+		newClaims["act"] = actClaim
+	} else if subjectActor != nil {
+		newClaims["act"] = subjectActor
+	}
+
+	for key, value := range userAttributes {
+		newClaims[key] = value
+	}
+
+	return newClaims
+}
+
+// validateAndExtractClaims validates a token and extracts its claims.
+func (h *tokenExchangeGrantHandler) validateAndExtractClaims(token string) (map[string]interface{}, error) {
+	// TODO: Validate for external token issuers
+	if err := h.jwtService.VerifyJWTSignature(token); err != nil {
```

Comment:
Shouldn't we validate the issuer is ours before verifying using Thunder key?

We should Ideally fail the flow if other issuers are not supported with the current implementation with proper error message rather than trying to validate with Thunder key and failing the validation.